### PR TITLE
Magento2 722 relations

### PR DESCRIPTION
**Pull request by sofib**
# Pull Request Template

## Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

## Type of change

- [x] Bug fix
- [ ] New feature
- [ ] This change requires a documentation update

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I updated the CHANGELOG.md

### CODE REVIEW

File: src/Model/Export/Product.php [hunk 1]
```diff
@@ -530,6 +582,7 @@ public function setRelations(): void
         parent::setRelations($result);
     }
 
+
```
Comment (Stereck) [inline src/Model/Export/Product.php]:
```suggestion
```

File: src/Model/Export/Product.php [hunk 2]
```diff
@@ -498,28 +499,79 @@ public function setTags(): void
         parent::setTags($result);
     }
 
+    private function isProductTypeSupported($typeInstance)
```
Comment (Stereck) [inline src/Model/Export/Product.php]: Please always add PHP doc for at least methods that have parameters or return values.

File: src/Model/Export/Product.php [hunk 3]
```diff
@@ -498,28 +499,79 @@ public function setTags(): void
...
+    {
+        if ($typeInstance instanceof Simple) {
```
Comment (Stereck) [inline src/Model/Export/Product.php]: No must have, but maybe a switch case would be easier to read here.

File: src/Model/Export/Product.php [hunk 4]
```diff
@@ -498,28 +499,79 @@ public function setTags(): void
...
+        $relatedProducts = $this->item->getRelatedProducts();
+
+        if (empty($crossSell) && empty($upsell) && empty($relatedProducts)) {
+            parent::setRelations([]);
+            return;
+        }
+
+        // in order to avoid making to many sql requests here we reaggregate all relations and decompose them at the end
+
+        $result = [];
```
Comment (Stereck) [inline src/Model/Export/Product.php]: Please user alignment of = in this project.

File: src/Model/Export/Product.php [hunk 5]
```diff
@@ -498,28 +499,79 @@ public function setTags(): void
...
+        $relation = new Product\Relation($crossSell, $upsell, $relatedProducts);
+
+        if ($relation->hasUnprocessedRelations()) {
+            // check configurable
+            $objectManager = \Magento\Framework\App\ObjectManager::getInstance();
+            $rows = $objectManager->create('Shopgate\Export\Model\ResourceModel\ConfigurableProduct')->getParentAndChildIdsByChildIds($relation->getUnprocessedRelationIds());
```
Comment (Stereck) [inline src/Model/Export/Product.php]: Can you please break those long lines?

File: src/Model/Export/Product.php (hunk 2 above)
Comment (Stereck) [reply src/Model/Export/Product.php]: This method is a duplicate and seemingly never called.

**Review by Stereck — CHANGES_REQUESTED** [review]
Just from gut feeling this adds much more complexity as expected, but I am not into the topic to judge if this can be simplified. However code in general looks fine except some CS adjustments that are needed to match project CS.

File: src/Model/Export/Product.php (hunk 3 above)
Comment (sofib) [reply src/Model/Export/Product.php]: I think that's very individual - it is just 4 code paths here.

File: src/Model/Export/Product.php (hunk 2 above)
Comment (sofib) [reply src/Model/Export/Product.php]: Right

File: src/Model/Export/Product.php (hunk 5 above)
Comment (sofib) [reply src/Model/Export/Product.php]: yes

File: src/Model/Export/Product.php (hunk 4 above)
Comment (sofib) [reply src/Model/Export/Product.php]: Unfortunately we have no auto formatter for vs code or via ci. Done

**Comment by sofib** [comment]
> Just from gut feeling this adds much more complexity as expected, but I am not into the topic to judge if this can be simplified. However code in general looks fine except some CS adjustments that are needed to match project CS.

There was an effort on reducing/limiting number of db queries that would otherwise happen if we'd query for each relation of a product separately. Export is already slowish but I still also can't be a judge of that.

File: src/Model/Export/Product.php [hunk 6]
```diff
@@ -503,23 +503,59 @@ public function setTags(): void
...
+        }
+
+        if ($relation->hasUnprocessedRelations()) {
+            // all the relations not resolved, we put as they so far were (direct id -> uid)
+            $relation->processRemainingRelationsAsDirectLinks();
+        }
+
+        $crossSellIds = $relation->getCrossSellIds();
+        $upsellIds = $relation->getUpsellIds();
+        $relatedProductIds = $relation->getRelatedProductIds();
```
Comment (Stereck) [inline src/Model/Export/Product.php]:
```suggestion
        $crossSellIds      = $relation->getCrossSellIds();
        $upsellIds         = $relation->getUpsellIds();
        $relatedProductIds = $relation->getRelatedProductIds();
```

File: src/Model/Export/Product/Relation.php [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+     *
+     * @param  array $products
+     * @param  string $classifier
+     * @return array
+     */
+    private function extractRelations(array $products, string $classifier) : array
+    {
+        $relationIds = [];
+        $relationsLookup = [];
+        $classifierLookup = [];
```
Comment (Stereck) [inline src/Model/Export/Product/Relation.php]:
```suggestion
        $relationIds      = [];
        $relationsLookup  = [];
        $classifierLookup = [];
```

**Review by Stereck — APPROVED** [review]

**Pull request merged**